README.md
# kancolle-auto

## IMPORTANT! Please see the [important announcement regarding the 2017 Fall Event ](https://github.com/mrmin123/kancolle-auto/issues/418) regarding ongoing support of kancolle-auto!

**kancolle-auto** is a robust Kantai Collection automation tool run from the command line. Please see the [kancolle-auto GUI APP (KAGA)](https://github.com/waicool20/KAGA), developed by waicool20, for a GUI for kancolle-auto (note: KAGA is not maintained by the developer of kancolle-auto).

***

>**WARNINGS/DISCLAIMERS**

> kancolle-auto is meant for educational purposes only. Actual and prolonged use of kancolle-auto may result in your account being banned. Remember that botting is against the rules!

> I make no guarantees that you will not be caught and penalized using kancolle-auto, so be smart about it. Don't spam expeditions and sorties nonstop 24/7. Try to mimic a human as much as possible with your use of this tool! Relevant discussion can be found [here](https://github.com/mrmin123/kancolle-auto/issues/130).

> In addition, if you let kancolle-auto sortie you might lose ships! It is highly unlikely (multiple checks occur to prevent this from happening) but I make no guarantees! If you're using a viewer with subtitles, please read the 1st question and answer of the [FAQ](#faqcommon-issues)! Also, if you let kancolle-auto use buckets, make sure you can spare them!

***

>**NOTE**

> kancolle-auto is **not** designed to be the **fastest** automation tool. It is designed to be **robust** and **highly customizable**. kancolle-auto is meant to free up your time, energy, and attention, not to net you the most resources or XP in the shortest time possible. It can automate almost every major feature in the game, including combat, and it can run for days on end with minimal to no user intervention.

***

**Please refer to the [releases page](https://github.com/mrmin123/kancolle-auto/releases) for stable releases, or the [master branch](https://github.com/mrmin123/kancolle-auto/archive/master.zip) for bleeding-edge (stable for the most part, but may potentially contain untested and/or buggy code in exchange for additional features). For Event support, please use the master branch for most up-to-date compatibility with the Event.**

Please read the [**kancolle-auto wiki**](https://github.com/mrmin123/kancolle-auto/wiki) for more details:

* The [**Changelog**](http://github.com/mrmin123/kancolle-auto/wiki/Changelog) will have information on the differences between the releases and master branch
* Please refer to the [**Setup**](https://github.com/mrmin123/kancolle-auto/wiki/Setup) page for instructions on setting up kancolle-auto
* Examples of the config can be found [**here**](https://github.com/mrmin123/kancolle-auto/wiki/Example-configs), while Event-specific config examples can be found [**here**](https://github.com/mrmin123/kancolle-auto/wiki/Event-configs)


### Features

* Expedition module &mdash; automate expeditions
* PvP module &mdash; automate PvP, including diamond and line abreast formations versus submarine-heavy fleets
* Combat module &mdash; automate sorties, node selections, repairs, and submarine switching
  * Supports sorties to Event maps, Combined Fleets, LBAS, and pre-boss/boss support expeditions
* Quests module &mdash; automate quests
* Individual toggles for each of the above modules
* Scheduled sleeping/pausing of script
* Rudimentary catbomb recovery
* Separate `config.ini` file for easier configuration and backup of configurations
* Random variations to help avoid bot detection
* Helpful timers and other messages in console (when using SikuliX in command line mode)

For a more in-depth list of features, as well as installation/usage directions, please refer to the [kancolle-auto wiki](http://github.com/mrmin123/kancolle-auto/wiki).

kancolle-auto was originally a fork of [these](https://github.com/amylase/kancolle-auto) [other](https://github.com/Yukariin/kancolle-auto) [projects/forks](https://github.com/kevin01523/kancolle-auto), but has since outgrown on them in scope and function. Some ideas were inspired by [another similiar tool](https://github.com/tantinevincent/Onegai-ooyodosan).

### Quick Start

1. Install [Java JRE 8](http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html)
2. Install the latest [SikuliX Nightly](http://nightly.sikuli.de/) with options 1 (Pack 1) and 3 (Tesseract OCR) selected (these are important!)
3. Install kancolle-auto
4. [Setup kancolle-auto's config file](https://github.com/mrmin123/kancolle-auto/wiki/Setup-config.ini) ([standard examples](https://github.com/mrmin123/kancolle-auto/wiki/Example-configs)/[event examples](https://github.com/mrmin123/kancolle-auto/wiki/Event-configs))
5. Run Kantai Collection
6. Run kancolle-auto using command `java -jar <path_to_sikuli>/sikulix.jar -r <path_to_kancolle_auto>/kancolle_auto.sikuli` (replacing `<path_to_sikuli>` and `<path_to_kancolle_auto>` with the correct directories for your installs)

### FAQ/Common Issues

#### Q: Can I lose my ships using kancolle-auto?

A: It's highly unlikely, but I can make no guarantees. There is a caveat to 'highly unlikely', however: if you're using a viewer with subtitles and those subtitles cover up ship portraits on the post-combat results screen (typically the last ship in a fleet if the subtitles are at the bottom of the screen), it can hinder kancolle-auto's ability to accurately detect ships in danger of being sunk. Either lower the subtitle's font size so that it does not cover up the results screen (I personally use a font size of 12px on KC3Kai), or disable them entirely.

You could also lose ships if you set the Combat module's LastNodePush to `True` and accidentally push into a comabt node. Hopefully the warnings in the config file will help you make the right decision regarding this flag.

#### Q: Can I do other stuff on the machine/play Kantai Collection while kancolle-auto is running?

A: Sometimes. kancolle-auto takes control of the mouse as it runs so it will be difficult to do anything meaningful while kancolle-auto is actively running. During scheduled pauses, however, you have full control of the machine. Just make sure that Kantai Collection is not in the middle of a PvP or Sortie when the schedule pause ends. The script will crash and you will have to restart the script.

#### Q: kancolle-auto crashes on startup (AttributeError ( 'NoneType' object has no attribute 'x' ) error)

A: Check/try the following:

* If you have multiple monitors, make sure Kantai Collection is running on the primary monitors. Sikuli can only find windows on the primary monitor.
* Make sure your Kantai Collection game window is scaled/resized to 100% (no resize). kancolle-auto only works if your game is in its original scale.
* Disable any extensions or plugins that might obstruct the Kantai Collection game window.
* The moment the script starts, manually click on the game window to switch focus to it.

#### Q: kancolle-auto periodically restarts and/or crashes!

A: kancolle-auto is fairly stable at this moment. Assuming your dependencies (Java, Sikuli, and Tesseract) and config are set up properly, crashes might be specific to your viewer, machine, internet connection, etc. To help mitigate these issues please turn to the `SleepModifier` parameter in the config. This value raises the buffer time kancolle-auto takes between steps, and can address most issues not caused by actual bugs in the code. Begin with 1 and gradually raise it to 10. If this does not eliminate or reduce the number of restarts and crashes, only then please open a detailed issue ticket.

Side note 1: If you are on Windows, also try disabling time synchronization on the OS level ([relevant information](https://answers.launchpad.net/sikuli/+question/194095)).

Side note 2: Certain VMs and VM setups have been known to cause issues with Sikuli image matching. If you are running kancolle-auto on a VM, try restarting your environment, or switching to a different VM.

#### Q: kancolle-auto does not take timers into consideration and/or I get OCR warnings (n0text warnings)

A: You're missing the tesseract OCR files. Please re-install Sikuli with the correct settings.

#### Q: I started getting catbombed frequently after I started using kancolle-auto!

A: You probably botted too much and triggered the game's bot protection. Use the ScheduledSleep functionality and let the program pause itself for a few hours every day.

#### Q: This command line stuff blows... isn't there an easier and prettier way to run kancolle-auto?

A: Please refer to the [kancolle-auto GUI APP (KAGA)](https://github.com/waicool20/KAGA) for your GUI needs.


CHANGELOG.md
#### 2017-12-09 ([Release 10.4 Final](https://github.com/mrmin123/kancolle-auto/releases/tag/10.4))
* kancolle-auto officially retired. Please refer to [kcauto-kai](https://github.com/mrmin123/kcauto-kai) for your up-to-date automation needs. Join the [kcauto-kai Discord](https://discord.gg/smAhRKw) for announcements, discussion, and help.

#### 2017-10-25 ([Release 10.3](https://github.com/mrmin123/kancolle-auto/releases/tag/10.3))
* Update assets for Kantai Collection UI update and other compatibility

#### 2017-10-18 ([Release 10.2](https://github.com/mrmin123/kancolle-auto/releases/tag/10.2))
* Update assets and code for Kantai Collection UI Update
* Update pattern generator

#### 2017-09-29 ([Release 10.1](https://github.com/mrmin123/kancolle-auto/releases/tag/10.1))
* Add support for UIT-25 and I-504 (thanks to @twisting2017)
* Restore accidentally-removed combat assets

#### 2017-09-22
* Archive 2017 Summer Event assets

#### 2017-09-04
* Add support for 2017 Summer Event (thanks to @muromachi, @waicool20, @yhn147, @perryhuynh)
* Add support for Luigi Torelli in the submarine switcher module (thanks to @waicool20)

#### 2017-06-07 ([Release 10](https://github.com/mrmin123/kancolle-auto/releases/tag/10))
* Update PvP and Submarine Switch assets to support the June 6th 2017 update to Kantai Collection (thanks to everyone for reporting changes and providing assets)

#### 2017-05-31
* Actually archive 2017 Spring Event assets
* Change `FleetComp` option to `FleetComps` to support fleet cycling for sorties (thanks to @waicool20)
* Add `UseBuckets` option to Submarine Switch module to specify whether or not submarines should be bucketed (thanks to @R-Jimenez)
* Update readme

#### 2017-05-22
* Archive 2017 Spring Event assets

#### 2017-05-13
* Add additional assets for 2017 Spring Event (thanks to @frosty5689)

#### 2017-05-12
* Add support for 2017 Spring Event (thanks to @AppleBoi86, @waicool20, @twisting2017, @perryhuynh, and @frosty5689)

#### 2017-05-07
* Update asset for port checking for Events

#### 2017-04-03
* Bugfix to repair module (thanks to @Aripracc)

#### 2017-04-02
* Updates to sortie-related assets (thanks to @waicool20)

#### 2017-03-29 ([Release 9.1](https://github.com/mrmin123/kancolle-auto/releases/tag/9.1))
* Create wrapper for streamlining calls to Sikuli's `findAll` functionality

#### 2017-03-26
* Fix fleetcomp switcher and assets
* Update readme and configs

#### 2017-03-17 ([Release 9](https://github.com/mrmin123/kancolle-auto/releases/tag/9))
* Update assets to support the March 17th 2017 upgrade to Kantai Collection (thanks to everyone for reporting changes and providing assets)

#### 2017-03-15 ([Release 8.3](https://github.com/mrmin123/kancolle-auto/releases/tag/8.3))
* Bugfixes to PvP module

#### 2017-03-13 ([Release 8.2](https://github.com/mrmin123/kancolle-auto/releases/tag/8.2))
* Bugfixes to submarine switch module

#### 2017-03-12 ([Release 8.1](https://github.com/mrmin123/kancolle-auto/releases/tag/8.1))
* Impovements to PvP module so it selects diamond or line abreast vs sub-heavy opponents, as well as declining night battle against all-sub fleets
* Improvements to submarine switch module to use less clicks
* `EnabledSubs` option of submarine switch module now supports 'ss' and 'ssv' as valid group values
* Bugfix to `ScheduledStop` feature (thanks to @waicool20)
* Minor code updates

#### 2017-02-28 ([Release 8](https://github.com/mrmin123/kancolle-auto/releases/tag/8))
* Archive 2017 Winter Event assets
* Implement `FatigueSwitch` feature, which lets you specify whether or not kancolle-auto should attempt to switch out fatigued submarines with its submarine switcher module
* Optimized the submarine switch module's performance when `EnabledSubs` is set to 'all'
* Bugfix to fix non-detection of quest bw7
* Minor code updates

#### 2017-02-25
* Critical bugfix to `LastNodePush` feature. Please upgrade if you are using this, otherwise you are at risk of losing ships

#### 2017-02-24
* Bugfix with config reader not working when `ReplaceLimit` was set to blank value

#### 2017-02-23
* Implement `ReplaceLimit`, which lets you specify what up to what level of damaged subs should be swapped out by the submarine switch module
* Revise timings to fix bugs caused by optimizing default Sikuli wait times
* Bugfix on 2017 Winter Event asset (E-3 panel)

#### 2017-02-20
* Optimized default Sikuli and combat loop wait times
* Bugfix and updates to config files
* Revise some expedition assets

#### 2017-02-19
* Update 2017 Winter Event assets
* Add SSV support to submarine switch module (thanks to @waicool20)
* Revise submarine switch assets
* Improve `debug_find` function
* Add CONTRIBUTING.md
* Update config and config_detailed (may cause many merge conflicts)
* Update code to follow PEP8 spacing
* Copy 2017 Winter Event assets into archive folder for use in wiki documentation

#### 2017-02-18
* Bugfix to LBAS assignment method
* Bugfix to repair module's setting of next sortie time

#### 2017-02-17
* Expand support for 2017 Winter Event

#### 2017-02-16
* Expand support for 2017 Winter Event

#### 2017-02-15
* Expand support for 2017 Winter Event (thanks to @waicool20)
* Improve Event expedition support (thanks to @waicool20)

#### 2017-02-14
* Expand support for 2017 Winter Event (thanks to @AppleBoi86 and @Danielosama)

#### 2017-02-12
* Expand support for 2017 Winter Event
* Revise submarine switch module logic

#### 2017-02-11
* Begin adding support for 2017 Winter Event
* Bugfix to LBAS panel switch image so it works with low-morale planes

#### 2017-02-07 ([Release 7.1](https://github.com/mrmin123/kancolle-auto/releases/tag/7.1))
* Improvements to repair and submarine switch modules (thanks to @waicool20)
* Further modifications to combat module
* Minor updates and code cleanup

#### 2017-01-28
* Bugfix for crash when kancolle-auto's repair module encounters a completely filled repair screen (thanks to @waicool20)

#### 2017-01-23

kancolle_auto.sikuli/config_reader.sikuli/config_reader.py
# Config-reader module
import ConfigParser, sys
from sikuli import *
from util import *


# Load the config.ini file
def get_config(settings, sleep_cycle):
    log_msg("Reading config file!")
    # Change paths and read config.ini
    os.chdir(getBundlePath())
    os.chdir('..')
    config = ConfigParser.ConfigParser()
    config.read('config.ini')
    # Set user settings
    # 'General' section
    settings['program'] = config.get('General', 'Program')
    settings['recovery_method'] = config.get('General', 'RecoveryMethod')
    settings['basic_recovery'] = config.getboolean('General', 'BasicRecovery')
    settings['jst_offset'] = config.getint('General', 'JSTOffset')
    sleep_cycle = config.getint('General', 'SleepCycle')
    # 'Scheduled Sleep' section
    if config.getboolean('ScheduledSleep', 'Enabled'):
        settings['scheduled_sleep_enabled'] = True
        settings['scheduled_sleep_start'] = "%04d" % config.getint('ScheduledSleep', 'StartTime')
        settings['scheduled_sleep_length'] = config.getfloat('ScheduledSleep', 'SleepLength')
        log_msg("Scheduled Sleep enabled: ~%s hours starting at ~%s" % (settings['scheduled_sleep_length'], settings['scheduled_sleep_start']))
    else:
        settings['scheduled_sleep_enabled'] = False
    # 'Scheduled Stop' section
    if config.getboolean('ScheduledStop', 'Enabled'):
        settings['scheduled_stop_enabled'] = True
        settings['scheduled_stop_mode'] = config.get('ScheduledStop', 'Mode').lower()
        if (settings['scheduled_stop_mode'] not in ['time', 'expedition', 'sortie', 'pvp']):
            log_error("'%s' is not a scheduled stop mode! Please check your config file." % settings['scheduled_stop_mode'])
            sys.exit()
        settings['scheduled_stop_count'] = config.getint('ScheduledStop', 'Count')
        log_msg("Scheduled Stop enabled: %s (%s)" % (settings['scheduled_stop_count'], settings['scheduled_stop_mode']))
    else:
        settings['scheduled_stop_enabled'] = False
    # 'Expeditions' section
    if config.getboolean('Expeditions', 'Enabled'):
        settings['expeditions_enabled'] = True
        if config.get('Expeditions', 'Fleet2'):
            settings['expedition_id_fleet_map'][2] = config.getint('Expeditions', 'Fleet2')
        if config.get('Expeditions', 'Fleet3'):
            settings['expedition_id_fleet_map'][3] = config.getint('Expeditions', 'Fleet3')
        if config.get('Expeditions', 'Fleet4'):
            settings['expedition_id_fleet_map'][4] = config.getint('Expeditions', 'Fleet4')
        log_msg("Expeditions enabled: %s" % (', '.join('fleet %s: %s' % (key, settings['expedition_id_fleet_map'][key]) for key in sorted(settings['expedition_id_fleet_map'].keys()))))
    else:
        settings['expeditions_enabled'] = False
    # 'PvP' section
    if config.getboolean('PvP', 'Enabled'):
        settings['pvp_enabled'] = True
        settings['pvp_fleetcomp'] = config.getint('PvP', 'FleetComp')
        log_msg("Combat enabled (PvP mode)")
    else:
        settings['pvp_enabled'] = False
    # 'Combat' section
    if config.getboolean('Combat', 'Enabled'):
        settings['combat_enabled'] = True
        settings['combat_fleetcomps'] = []
        for fleet in filter(None, config.get('Combat', 'FleetComps').replace(' ', '').split(',')):
            try:
                fleet_num = int(fleet)
                if 1 <= fleet_num <= 10:
                    settings['combat_fleetcomps'].append(fleet_num)
                else:
                    raise ValueError("Value %s must be a number between 1 - 10 inclusive" % fleet)
            except ValueError, e:
                log_error(str(e))
                log_error("Invalid value found: %s" % fleet)
                sys.exit(1)
        if len(settings['combat_fleetcomps']) > 1:
            log_success("Combat module cycling fleetcomps %s" % ', '.join(map(str, settings['combat_fleetcomps'])))
        settings['combat_area'] = config.get('Combat', 'Area')
        settings['combat_subarea'] = config.get('Combat', 'Subarea')
        settings['combined_fleet'] = config.getboolean('Combat', 'CombinedFleet')
        if settings['combined_fleet']:
            # Remove fleet 2 from expedition list if combined fleet is enabled
            settings['expedition_id_fleet_map'].pop(2, None)
            # Disable PvP if combined fleet is enabled
            settings['pvp_enabled'] = False
            settings_check_valid_formations = ['combinedfleet_1', 'combinedfleet_2', 'combinedfleet_3', 'combinedfleet_4']
            settings_check_valid_formations += ['line_ahead', 'double_line', 'diamond', 'echelon', 'line_abreast', ]
            settings_check_filler_formation = 'combinedfleet_4'
        else:
            settings_check_valid_formations = ['line_ahead', 'double_line', 'diamond', 'echelon', 'line_abreast', ]
            settings_check_filler_formation = 'line_ahead'
        settings['nodes'] = config.getint('Combat', 'Nodes')
        settings['node_selects'] = config.get('Combat', 'NodeSelects').replace(' ', '').split(',')
        if '' in settings['node_selects']:
            settings['node_selects'].remove('')
        settings['formations'] = config.get('Combat', 'Formations').replace(' ', '').split(',')
        # Check that supplied formations are valid
        for formation in settings['formations']:
            if formation not in settings_check_valid_formations:
                log_error("'%s' is not a valid formation! Please check your config file." % formation)
                sys.exit()
        if len(settings['formations']) < settings['nodes']:
            settings['formations'].extend([settings_check_filler_formation] * (settings['nodes'] - len(settings['formations'])))
        settings['night_battles'] = config.get('Combat', 'NightBattles').replace(' ', '').split(',')
        if len(settings['night_battles']) < settings['nodes']:
            settings['night_battles'].extend(['True'] * (settings['nodes'] - len(settings['night_battles'])))
        settings['retreat_limit'] = config.getint('Combat', 'RetreatLimit')
        settings['repair_limit'] = config.getint('Combat', 'RepairLimit')
        settings['repair_time_limit'] = config.getint('Combat', 'RepairTimeLimit')
        settings['reserve_docks'] = config.getboolean('Combat', 'ReserveDocks')
        settings['check_fatigue'] = config.getboolean('Combat', 'CheckFatigue')
        settings['port_check'] = config.getboolean('Combat', 'PortCheck')
        settings['medal_stop'] = config.getboolean('Combat', 'MedalStop')
        settings['last_node_push'] = config.getboolean('Combat', 'LastNodePush')
        log_msg("Combat enabled (sortie mode): sortieing to %s-%s" % (settings['combat_area'], settings['combat_subarea']))
    else:
        settings['combat_enabled'] = False
    # 'SubmarineSwitch' section
    if config.getboolean('SubmarineSwitch', 'Enabled') and settings['combat_enabled']:
        settings['submarine_switch'] = True
        settings_check_valid_subs = ['all', 'ss', 'ssv', 'i-8', 'i-8-kai', 'i-13', 'i-14', 'i-19', 'i-19-kai', 'i-26', 'i-26-kai', 'i-58', 'i-58-kai', 'i-168', 'i-401', 'maruyu', 'ro-500', 'u-511', 'luigi', 'uit-25', 'i-504']
        settings['submarine_switch_subs'] = config.get('SubmarineSwitch', 'EnabledSubs').replace(' ', '').lower().split(',')
        settings['submarine_switch_fatigue_switch'] = config.getboolean('SubmarineSwitch', 'FatigueSwitch')
        settings['submarine_switch_use_buckets'] = config.getboolean('SubmarineSwitch', 'UseBuckets')
        try:
            settings['submarine_switch_replace_limit'] = config.getint('SubmarineSwitch', 'ReplaceLimit')
        except ValueError:
            settings['submarine_switch_replace_limit'] = None
        # If 'all' is specified, disregard the other specified options
        if 'all' in settings['submarine_switch_subs']:
            settings['submarine_switch_subs'] = ['i-8', 'i-8-kai', 'i-13', 'i-14', 'i-19', 'i-19-kai', 'i-26', 'i-26-kai', 'i-58', 'i-58-kai', 'i-168', 'i-401', 'maruyu', 'ro-500', 'u-511', 'luigi', 'uit-25', 'i-504']
        else:
            # Expand 'ss' and 'ssv' group values to their actual subs
            if 'ss' in settings['submarine_switch_subs']:
                settings['submarine_switch_subs'].remove('ss')
                settings['submarine_switch_subs'].extend(['i-8', 'i-19', 'i-26', 'i-58', 'i-168', 'maruyu', 'ro-500', 'u-511', 'luigi', 'uit-25', 'i-504'])
            if 'ssv' in settings['submarine_switch_subs']:
                settings['submarine_switch_subs'].remove('ssv')
                settings['submarine_switch_subs'].extend(['i-8-kai', 'i-13', 'i-14', 'i-19-kai', 'i-26-kai', 'i-58-kai', 'i-401'])
            settings['submarine_switch_subs'] = list(set(settings['submarine_switch_subs']))
        for sub in settings['submarine_switch_subs']:
            if sub not in settings_check_valid_subs:
                log_error("'%s' is not a valid sub selection! Please check your config file." % sub)
                sys.exit()
        log_msg("Submarine Switch enabled")
    else:
        settings['submarine_switch'] = False
    # 'LBAS' section
    if config.getboolean('LBAS', 'Enabled') and settings['combat_enabled']:
        settings['lbas_enabled'] = True
        settings['lbas_groups'] = config.get('LBAS', 'EnabledGroups').replace(' ', '').split(',')
        if (len(settings['lbas_groups']) < 1):
            log_error("If LBAS is enabled, you must specify at least one active LBAS group")
            sys.exit()
        else:
            for i, val in enumerate(settings['lbas_groups']):
                settings['lbas_groups'][i] = int(val)
        settings['lbas_groups'].sort()
        settings['lbas_group_1_nodes'] = config.get('LBAS', 'Group1Nodes').replace(' ', '').split(',')
        if (1 in settings['lbas_groups'] and (settings['lbas_group_1_nodes'] != [''] and len(settings['lbas_group_1_nodes']) != 2)):
            log_error("You must specify zero (0) or two (2) nodes for active LBAS group 1!")
            sys.exit()
        settings['lbas_group_2_nodes'] = config.get('LBAS', 'Group2Nodes').replace(' ', '').split(',')
        if (2 in settings['lbas_groups'] and (settings['lbas_group_2_nodes'] != [''] and len(settings['lbas_group_2_nodes']) != 2)):
            log_error("You must specify zero (0) or two (2) nodes for active LBAS group 2!")
            sys.exit()
        settings['lbas_group_3_nodes'] = config.get('LBAS', 'Group3Nodes').replace(' ', '').split(',')
        if (3 in settings['lbas_groups'] and (settings['lbas_group_3_nodes'] != [''] and len(settings['lbas_group_3_nodes']) != 2)):
            log_error("You must specify zero (0) or two (2) nodes for active LBAS group 3!")
            sys.exit()
        log_msg("LBAS enabled: groups %s" % ', '.join(str(group) for group in settings['lbas_groups']))
    else:
        settings['lbas_enabled'] = False
    # 'Quests' section
    settings['active_quests'] = config.get('Quests', 'Quests').replace(' ', '').split(',')
    settings['active_quests'].sort()
    if config.getboolean('Quests', 'Enabled') and len(settings['active_quests']) > 0:
        settings['quests_enabled'] = True
        settings['quests_check_schedule'] = config.getint('Quests', 'CheckSchedule')
        log_msg("Quests enabled")
    else:
        settings['quests_enabled'] = False
    return settings, sleep_cycle
